### ISSUE TITLE
reshuffle_each_iteration=False ignored on validation (tf.dataset + Keras)

### ISSUE THREAD

**Issue by Kerybas** [tensorflow/tensorflow#37045]
When using .shuffle on some `tf.Dataset.from_generator`, it appears that the `reshuffle_each_iteration=False` is ignored by Keras.  
If you run the following example, you would expect to see the train_set and val_set buffer filling at the start of the session, and then you would no longer see it between each epoch. The buffers are large enough to cover more than one epoch and can be refilled in parallel during the epoch.    

However, what seems to happen is that the val_set shuffle buffer is filled at the end of each epoch.  
The terminal shows "Filling shuffle buffer" at the end of each epoch.  
You can see that it is specifically the val_set buffer by removing the shuffle on val_set. Doing so, you will no longer see a "Filling shuffle" at each epoch, only at the start of the session.  

Additionally, [issue 35100](https://github.com/tensorflow/tensorflow/issues/35100 ) will spam the terminal on this example, with or without shuffle.  

Seen on Win10, with tensorflow 2.10.  
Code to reproduce the problem:  
```
import tensorflow as tf
import numpy as np
import time

def random_gen():
    while True:
        time.sleep(0.01)
        yield (np.random.rand(1,), np.random.rand(1,))

dataset = tf.data.Dataset.from_generator(
        random_gen,
        (tf.float32, tf.float32),
        (tf.TensorShape((None,)), tf.TensorShape((None,))),
    )

train_set =   (dataset
               .shuffle(2000, reshuffle_each_iteration=False)
               .batch(32))
val_set =    (dataset
               .shuffle(2000, reshuffle_each_iteration=False) # Problem is here
               .batch(32))

model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(3))
model.compile(loss='mse', optimizer='adam')

model.fit(x=train_set,
          epochs= 5,
          steps_per_epoch= 10,
          validation_data= val_set,
          validation_steps= 5,
)
```

**Comment by ravikyram** [comment]
@Kerybas 
I have tried on colab with TF version 2.1.0 on val_set with `reshuffle_each_iteration=False` and `reshuffle_each_iteration=True`. Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/3523fea5be768181e2950328745819cf/untitled674.ipynb). Is this the expected behavior? Thanks!

**Comment by Kerybas** [comment]
@ravikyram 
The issue won't show up on a Notebook. I don't know why but Notebooks do not show the buffer filling outputs that you get in the terminal.

**Comment by mayankatsrijan** [comment]
Hi, I'm facing the same issue when I'm running my training module via terminal. Even though the `reshuffle_each_iteration=False`, In each epoch, it starts filling the shuffle buffer due to which training time is getting highly affected. Any update on this to how to overcome this?

**Comment by mk9440** [comment]
Hi, I'm facing the same issue when I'm running my training module via terminal. Even though the `reshuffle_each_iteration=False`, In each epoch, it starts filling the shuffle buffer due to which training time is getting highly affected. Any update on this to how to overcome this?

**Comment by jsimsa** [comment]
`reshuffle_each_iteration` does not control whether to shuffle each iteration or not. It controls whether different order of shuffling should be used.

If your input pipeline contains `shuffle`, then each epoch will perform shuffling. If you wish to overlap filling up of shuffle buffer with computation, you should put `repeat` after `shuffle`. This will result in the filling up of the shuffle buffer needed for the 2nd epoch to be amortized over the 1st epoch (as opposed to happening at the beginning of each epoch) and so on. You might also want to put `prefetch` at the end of your input pipeline to overlap training computation with preprocessing computation (see [tf.data performance guide](https://www.tensorflow.org/guide/data_performance) for more details).

**Comment by DamRsn** [comment]
I'm also experiencing this issue. 
I'm training a keras model with `model.fit` and `steps_per_epoch` argument set to a way lower value than the actual size of the dataset, and the dataset gets shuffled at each model epoch (while `reshuffle_each_iteration = False`). I can see that because when I'm using a large shuffle buffer size, a message is printed in the terminal stating that the buffer is currently being filled (and that it may take a while ...). Also I'm already using `repeat(-1)` and `prefetch` when I create the dataset. 
On my side, it seems to be the training dataset that gets shuffled and not the validation one as it is cached. 

I would need the dataset to be reshuffled only when it has been completely iterated through and not when I'm starting a new model epoch (that I defined to be n steps with `steps_per_epoch`).

Is there a way to do this and still use `model.fit` ? If it is not the case, I think it would be a very useful feature to have, to be able to disconnect "dataset epochs" (when the dataset have completely iterated through) and "model epochs" (after n_steps_per_epoch in `model.fit`), to be able to shuffle only after each "dataset epochs" .

The only solution I see now is to abandon `model.fit` and create a custom training loop receiving data from an iterator on the `tf.Dataset`.

In case it is needed, I'm using tensorflow 2.2.0 installed with anaconda (build: gpu_py37h1a511ff_0) on Ubuntu 20.04.

**Comment by sachinprasadhs** [comment]
While reproducing your issue in Tf Nightly, faced different error, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/75b97702d11b3b800d617e622acea69a/35650.ipynb). Thanks!
